### <div style="text-align: right"> RNA-sequencing </div> 

## **Differential Gene Expression Detection from Bulk RNA-Sequencing Data in Breast Cancer Subtypes**

### **Overview**
This repository contains scripts and analyses needed for processing RNA-Seq data from different breast cancer subtypes (HER2+, TNBC, and non-TNBC) compared to healthy samples (Normal). The aim of this project is to identify differentially expressed genes (DEGs) and enriched biological pathways that may contribute to cancer development and progression.<br>
The workflow consists of quality control, read mapping, gene expression quantification, differential expression analysis, and overrepresentation analysis (functional enrichment analysis).

### **Repository structure (steps to follow, order of scripts):**
* **01_run_FastQC.sh**
* **02_get_reference_annotation.sh**
* **03_do_unzip_fa_gtf.sh**
* **04_run_HISAT2_indexing.sh**
* **05_do_mapping_convert.sh**
* **06_run_samtools_sort.sh**
* **07_run_samtools_indexing.sh**
* **08_run_featureCounts.sh**
* **09_reformat_featureCounts.sh**
* **10_exploratory_DE_overrepresentation_analysis.R**

### This will result in the following files (outputs from each respective script):

| **Step** | **Output** | **Description** |
| ---- | ------------------------- | ---------------------------- |
| 01 | *_fastqc.html | Quality control report generated by FastQC. |
| 02 | Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz | Compressed FASTA file containing the reference genome sequence for human. |
|    | Homo_sapiens.GRCh38.113.gtf.gz | Compressed GTF file containing the gene annotation for human. |
| 03 | genome.fa | Unzipped FASTA file containing the human reference genome sequence.  |
|    | homo_sapiens.gtf | Unzipped GTF file containing the gene annotations for human. |
| 04 | genome_index.* | HISAT2 index files generated for the reference genome. |
| 05 | sdfsdf*_hisat2_summary.logsdfsdf | A log file for each read pair containing HISAT2 mapping statistics (e.g., mapping rate, alignments, etc.). |
|    | *_mappedReads.bam | A BAM file for each read pair (R1 and R2). |
| 06 | *mappedReads_sorted.bam | The sorted BAM file for each input BAM file. |
|    | sort_success.log | A log file documenting successful sorting operations, with timestamps and the associated file names. |
|    | *mappedReads_sorted.bam | The sorted BAM file for each input BAM file. |
| 07 | *mappedReads_sorted.bam.bai | The BAM index file (.bai) created for each sorted BAM file. |
| 08 | gene_counts.txt | A text file containing the read counts for each gene.  |
| 09 | reformatted_counts.txt | A text file with the reformatted count data from gene_counts.txt, where the first column contains gene IDs and the remaining columns contain read counts from the samples. No extra information like in gene_counts.txt. Input for Analysis in R (Step 10). |
| 10 | PCA_plot.png | PCA plot to assess how samples cluster based on their gene expression profiles. |
|    | Heatmap_sample_distance.png | Sample distance heatmap to visualize the pairwise distances or similarities between samples. |
|    | Volcano_Normal_vs_HER2.png | Volcano plot of HER2+ vs Normal, highlighting significant DEGs by plotting the log2FoldChange against p-values for each comparison. |
|    | Volcano_Normal_vs_NonTNBC.png | Volcano plot of non-TNBC vs Normal, highlighting significant DEGs by plotting the log2FoldChange against p-values for each comparison. |
|    | Volcano_Normal_vs_TNBC.png | Volcano plot of TNBC vs Normal, highlighting significant DEGs by plotting the log2FoldChange against p-values for each comparison. |
|    | Heatmap_gene_expression.png | Heatmap visualizing the expression patterns of the selected genes (based on the original publication) across all conditions. |
|    | VennDiagram_DE_genes.png |  visualize the overlapping and unique DEGs across the three conditions (comparisons). |
|    | GO_enrichment_her2.tsv | GO results of HER2+ vs Normal as a TSV file. |
|    | GO_enrichment_nontnbc.tsv | GO results of non-TNBC vs Normal as a TSV file. |
|    | GO_enrichment_tnbc.tsv | GO results of TNBC vs Normal as a TSV file. |
|    | GO_enrichment_her2_barplot.png | Barplot of enriched GO pathways in HER2+ vs Normal. |
|    | GO_enrichment_nontnbc_barplot.png | Barplot of enriched GO pathways in non-TNBC vs Normal. |
|    | GO_enrichment_tnbc_barplot.png | Barplot of enriched GO pathways in TNBC vs Normal. |
|    | GO_enrichment_her2_dotplot.png | Dot plot of enriched GO pathways in HER2+ vs Normal. |
|    | GO_enrichment_nontnbc_dotplot.png | Dot plot of enriched GO pathways in non-TNBC vs Normal. |
|    | GO_enrichment_tnbc_dotplot.png |  Dot plot of enriched GO pathways in TNBC vs Normal. |
|    | Heatmap_top_genes_expression.png |  Heatmap of gene expression of the top two genes from the top 3 GO pathways for all comparisons. |

### Note - Optional:
The repository includes an optional trimming script "optional_run_TrimFastQC.sh", which was not utilized in this analysis as the quality of the reads was deemed satisfactory.

### Required Tools:
To run the scripts, the following tools are necessary (version numbers are the version I used):
* **For shell scripts**: FastQC (v0.12.1), HISAT2 (v2.2.1), Samtools (v1.20),  featureCounts (v2.0.1)
* **For R scripts**: R (v4.4.2), DESeq2 (v1.46.0), ggplot2 (v3.5.1), pheatmap (v1.0.12), RColorBrewer (v1.1-3), EnhancedVolcano (v1.24.0), VennDiagram (v1.7.3), clusterProfiler (v4.14.4), org.Hs.eg.db (v3.20.0), DOSE (v4.0.0), enrichplot (v1.26.6)